sstables/sstable_set.cc

```diff
@@ -769,6 +770,393 @@ sstable_set_impl::create_single_key_sstable_reader(
     return make_combined_reader(schema, std::move(permit), std::move(readers), fwd, fwd_mr);
 }
 
+static logging::logger cormlogger("clustering_order_reader_merger");
+
+// Merges output from readers of TWCS sstables opened for a single partition query into a non-decreasing stream
+// of mutation fragments.
+//
+// Uses `min_position_reader_queue` to open the readers lazily as the read progresses through the partition.
+// A reader is opened only if we find that it may contain fragments for the currently inspected positions.
+//
+// Readers are closed as soon as we find that they were exhausted for the given partition query.
+//
+// Implements the `FragmentProducer` concept. However, `next_partition` and `fast_forward_to(partition_range)`
+// are not implemented and throw an error; the reader is only used for single partition queries.
+//
+// Assumes that the readers returned from `min_position_reader_queue` return fragments from the same partition.
+// Otherwise the behavior is unspecified.
+class clustering_order_reader_merger {
+    const schema_ptr _schema;
+    const reader_permit _permit;
+
+    // The key of the partition being queried.
+    const dht::decorated_key _key;
+
+    // Compares positions using *_schema.
+    const position_in_partition::tri_compare _cmp;
+
+    // A queue of readers used to lazily open new readers as we progress through the partition.
+    // Before a batch with position `p` is returned, we ensure that all readers containing positions
+    // <= `p` are popped from the queue.
+    min_position_reader_queue _reader_queue;
+
+    // Owning container for the readers popped from _reader_queue.
+    // If we are sure that a reader is exhausted (all rows from the queried partition have been returned),
+    // we destroy and remove it from the container.
+    std::list<reader_and_positions> _all_readers;
+    using reader_iterator = std::list<reader_and_positions>::iterator;
+
+    // A min-heap of readers, sorted by the positions of their next fragments.
+    // The iterators point to _all_readers.
+    // Invariant: every reader in `_peeked_readers` satisfies `!is_buffer_empty()`,
+    // so it is safe to call `pop_mutation_fragment()` and `peek_buffer()` on it.
+    merger_vector<reader_iterator> _peeked_readers;
```

Comment:
Where is this defined?

Reply:
You mean `merger_vector`? In merging_reader.hh